src/device/socket/vsock.rs

```diff
@@ -101,24 +139,185 @@ impl<'a, H: Hal, T: Transport> VirtIOSocket<'a, H, T> {
+        }
+    }
+
+    /// Shuts down the connection.
+    /// TODO: Add a timeout for shutdown.
+    pub fn shutdown(&mut self) -> Result {
+        let connection_info = self.connection_info()?;
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: connection_info.dst.cid.into(),
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::Shutdown.into(),
+            ..Default::default()
+        };
+        self.send_packet_to_tx_queue(&header, &[])?;
+        let received_header = self.pop_header_only_packet_from_rx_queue()?;
+        match received_header.op()? {
+            VirtioVsockOp::Rst | VirtioVsockOp::Shutdown => {
+                info!("Disconnected from the peer");
+                self.connection_info = None;
+                Ok(())
+            }
+            _ => Err(SocketError::InvalidOperation.into()),
+        }
+    }
+
+    /// Waits until there is at least one element to pop in rx queue.
+    fn wait_one_in_rx_queue(&self) {
+        for _ in 0..10000000 {
+            if self.rx.can_pop() {
+                break;
+            }
+        }
+    }
+
+    fn send_packet_to_tx_queue(&mut self, header: &VirtioVsockHdr, buffer: &[u8]) -> Result {
+        let _len = self.tx.add_notify_wait_pop(
+            &[header.as_bytes(), buffer],
+            &mut [],
+            &mut self.transport,
+        )?;
+        Ok(())
+    }
+
+    fn pop_packet_from_rx_queue(&mut self) -> Result<VirtioVsockPacket> {
         let token = if let Some(token) = self.rx.peek_used() {
             token // TODO: Use let else after updating Rust
         } else {
             return Err(SocketError::NoResponseReceived.into());
         };
+        let buffer_range = self.get_rx_buffer_range(token as usize)?;
+        let buffer = &mut self.rx_buf[buffer_range];
         // Safe because we are passing the same buffer as we passed to `VirtQueue::add`.
-        let _len = unsafe { self.rx.pop_used(token, &[], &mut [self.queue_buf_rx])? };
-        let packet_rx = VirtioVsockPacket::read_from(self.queue_buf_rx)?;
-        trace!("Received packet {:?}. Op {:?}", packet_rx, packet_rx.op());
-        match packet_rx.op()? {
-            VirtioVsockOp::Response => Ok(()),
-            VirtioVsockOp::Rst => Err(SocketError::ConnectionFailed.into()),
-            VirtioVsockOp::Invalid => Err(SocketError::InvalidOperation.into()),
-            _ => todo!(),
+        let _len = unsafe { self.rx.pop_used(token, &[], &mut [buffer])? };
+        let packet = VirtioVsockPacket::read_from(buffer)?;
+        debug!("Received packet {:?}. Op {:?}", packet, packet.hdr.op());
+        Ok(packet)
+    }
+
+    fn pop_header_only_packet_from_rx_queue(&mut self) -> Result<VirtioVsockHdr> {
+        let packet = self.pop_packet_from_rx_queue()?;
+        assert_eq!(packet.data.len(), 0);
```

Comment:
This should return an error rather than panicking, as it indicates a misbehaving device rather than a bug in the driver.